### PR TITLE
Add Zero Bubble Pipeline Parallelism H1 Schedule

### PR DESCRIPTION
**Pull request by nvmdava**
This repository contains the zbh1 implementation of the paper [Zero Bubble Pipeline Parallelism](https://arxiv.org/abs/2401.10241) on Megatron-Deepspeed architecture. The scheduler optimizes runtime by splitting and delaying the operations related to parameter gradient, allowing for much fine grained control over schedule.

To enable the optimization, simply add `--enable-zbh1-pipeline` flag before running. As the implementation is done for DeepSpeed's scheduler, make sure that you are also using DeepSpeed.

The implementation makes two simple changes. First it allows Megatron's backward function to split the weight gradient calculations and put them in a queue. Second, when zero bubble H1 is enabled, it overwrites deepspeed's schedule and instruction maps to run the custom schedule.

For optimization, when the backward pass is followed by a weight pass, the custom schedule does not split them. Due to this, the summation of gradients for weights may have different order and possibly create small numerical differences. If numerical accuracy is needed, it can be done by replacing minimal codes in the custom schedule, but note that there is a noticable performance drop when using bf16.

The implementation has been checked and verified on various setups such as ZeRO Stage 1, activation recomputation, flash attention, tensor parallel, data parallel and bf16. By approximate measure, ~10% acceleration was observed when microbatch count is twice the number of pipeline stages.

test results

### CODE REVIEW

**Comment by nvmdava** [comment]
@microsoft-github-policy-service agree company="Sea"

**Comment by tjruwase** [comment]
@nvmdava, thanks for this awesome PR. Zero Bubble PP is a really cool idea. We really appreciate your contribution. How did you test this for correctness? Can you also provide some convergence screenshots showing correctness?

**Comment by nvmdava** [comment]
@tjruwase, thanks for your interest. I have provided additional testing details on accuracy:
* If ZBH1 schedule is not enabled, then it does not affect the numeric accuracy, i.e the loss matches the current Megatron-DeepSpeed results.
* If numerical accuracy is a must when running ZBH1 schedule, you can achieve that by replacing every `BackwardPass(id)` in `deepspeed_zbh1_schedule.py` with `BackwardOnlyPass(id)` and `WeightPass()`.
* Otherwise, over 10'000 steps, the maximum loss drift is negligible with 0.0123 as maximum difference at step 6428. At 10'000th step the loss is 4.6043 for the original schedule and 4.6025 for the ZBH1 schedule.
loss

These tests were done on 1.3B model with `pp_size=8, global_batch_size=16, batch_size=1` with [RedPajama Sample's c4_sample](https://huggingface.co/datasets/togethercomputer/RedPajama-Data-1T-Sample/blob/main/c4_sample.jsonl)  as training data.

**Comment by tjruwase** [comment]
@nvmdava, thanks for providing this explanation. I think these details should be documented more formally, so can you please move them into https://github.com/microsoft/Megatron-DeepSpeed/tree/main/examples_deepspeed. That folder contains documentation, some in tutorial form, of major features to help user understanding. Please browse a few of them to get an idea and then move all the useful information in this PR into a folder (e.g., `zero_bubble_pp`) for this feature. Thanks!

**Comment by tjruwase** [comment]
> * If numerical accuracy is a must when running ZBH1 schedule, you can achieve that by replacing every `BackwardPass(id)` in `deepspeed_zbh1_schedule.py` with `BackwardOnlyPass(id)` and `WeightPass()`.

Couldn't you provide a command line argument to control this? I think that would be more convenient for users. I think this is also an interesting topic to highlight the documentation/tutorial.

**Comment by ufotalent** [comment]
Hi @tjruwase, thanks for the constructive feedbacks!

I'm Xinyi, one of the authors of the zero bubble pipeline parallelism paper. Amar @nvmdava was my intern and he finished his internship with us and sailed to another adventure. So I'll take on this PR.

I've added some documentation to `examples_deepspeed/zero_bubble_pp`. Also I added a flag to control the numerical accuracy called `--enable-zbh1-exact-semantics`

Thanks!
